File: po/zadania/szybkosc-programow-java/DijkstraBenchmark.java
import java.io.BufferedInputStream;
import java.io.FileInputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Scanner;

public class DijkstraBenchmark {
    static final long INF = 4_000_000_000_000_000_000L;

    static int n, m, k;
    static long seed;

    static ArrayList<ArrayList<int[]>> g;
    static int[] sources;

    static void addEdge(int u, int v, int w) {
        g.get(u).add(new int[] {v, w});
        g.get(v).add(new int[] {u, w});
    }

    static void readGraph(String fileName) throws Exception {
        Scanner fs = new Scanner(new BufferedInputStream(new FileInputStream(fileName)));

        n = fs.nextInt();
        m = fs.nextInt();
        k = fs.nextInt();
        seed = fs.nextLong();

        g = new ArrayList<>();
        for (int i = 0; i < n; i++) g.add(new ArrayList<>());

        sources = new int[k];
        for (int i = 0; i < k; i++) sources[i] = fs.nextInt();

        for (int i = 0; i < m; i++) {
            int u = fs.nextInt();
            int v = fs.nextInt();
            int w = fs.nextInt();
            addEdge(u, v, w);
        }

        fs.close();
    }

    static long dijkstra(int s) {
        long[] dist = new long[n];
        Arrays.fill(dist, INF);

        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        dist[s] = 0;
        pq.add(new long[] {0, s});

        while (!pq.isEmpty()) {
            long[] top = pq.poll();
            long du = top[0];
            int u = (int) top[1];

            if (du != dist[u]) continue;

            for (int[] edge : g.get(u)) {
                int v = edge[0];
                int w = edge[1];
                if (dist[v] > du + w) {
                    dist[v] = du + w;
                    pq.add(new long[] {dist[v], v});
                }
            }
        }

        long sum = 0;
        for (long x : dist) sum += x;
        return sum;
    }

    public static void main(String[] args) throws Exception {
        String fileName = "graph.txt";
        if (args.length >= 1) fileName = args[0];
        readGraph(fileName);

        long start = System.nanoTime();

        long checksum = 0;
        for (int s : sources) checksum += dijkstra(s);

        long finish = System.nanoTime();
        double elapsedMs = (finish - start) / 1e6;

        System.out.println("algorithm=dijkstra");
        System.out.println("graph=undirected_sparse");
        System.out.println("input_file=" + fileName);
        System.out.println("vertices=" + n);
        System.out.println("undirected_edges=" + m);
        System.out.println("sources=" + k);
        System.out.println("seed=" + seed);
        System.out.println("checksum=" + checksum);
        System.out.printf("elapsed_ms=%.3f%n", elapsedMs);
    }
}

File: po/zadania/szybkosc-programow-java/dijkstra_benchmark.cpp
#include <chrono>
#include <cstdint>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <queue>
#include <utility>
#include <vector>
using namespace std;

const int64_t INF = 4'000'000'000'000'000'000LL;

int n, m, k;
uint32_t seed;

vector<vector<pair<int, int>>> g;
vector<int> sources;

void add_edge(int u, int v, int w) {
    g[u].push_back({v, w});
    g[v].push_back({u, w});
}

bool read_graph(const string &file_name) {
    ifstream fin(file_name);
    if (!fin) return false;

    fin >> n >> m >> k >> seed;
    if (!fin || n < 2 || m < n - 1 || k <= 0) return false;

    g.assign(n, {});
    sources.resize(k);

    for (int i = 0; i < k; i++) fin >> sources[i];
    for (int i = 0; i < m; i++) {
        int u, v, w;
        fin >> u >> v >> w;
        add_edge(u, v, w);
    }

    return (bool) fin;
}

int64_t dijkstra(int s) {
    vector<int64_t> dist(n, INF);
    priority_queue<pair<int64_t, int>, vector<pair<int64_t, int>>, greater<pair<int64_t, int>>> pq;

    dist[s] = 0;
    pq.push({0, s});

    while (!pq.empty()) {
        auto [du, u] = pq.top();
        pq.pop();

        if (du != dist[u]) continue;

        for (auto [v, w] : g[u]) {
            if (dist[v] > du + w) {
                dist[v] = du + w;
                pq.push({dist[v], v});
            }
        }
    }

    int64_t sum = 0;
    for (int64_t x : dist) sum += x;
    return sum;
}

int main(int argc, char **argv) {
    string file_name = "graph.txt";
    if (argc >= 2) file_name = argv[1];

    if (!read_graph(file_name)) {
        cout << "Bad input file\n";
        return 0;
    }

    auto start = chrono::steady_clock::now();

    int64_t checksum = 0;
    for (int s : sources) checksum += dijkstra(s);

    auto finish = chrono::steady_clock::now();
    double elapsed_ms = chrono::duration<double, milli>(finish - start).count();

    cout << "algorithm=dijkstra\n";
    cout << "graph=undirected_sparse\n";
    cout << "input_file=" << file_name << '\n';
    cout << "vertices=" << n << '\n';
    cout << "undirected_edges=" << m << '\n';
    cout << "sources=" << k << '\n';
    cout << "seed=" << seed << '\n';
    cout << "checksum=" << checksum << '\n';
    cout << fixed << setprecision(3);
    cout << "elapsed_ms=" << elapsed_ms << '\n';

    return 0;
}
